sklearn/cluster/k_means_.py

```diff
@@ -685,7 +685,9 @@ def _init_centroids(X, k, init, random_state=None, x_squared_norms=None,
         seeds = random_state.permutation(n_samples)[:k]
         centers = X[seeds]
     elif hasattr(init, '__array__'):
-        centers = init
+        # ensure that the centers have the same dtype as X
+        # this is a requirement of fused types of cython
+        centers = np.array(init, dtype=X.dtype)
```

Comment:
I've added a test to make sure that we do copy `init` [here](https://github.com/scikit-learn/scikit-learn/pull/6846/commits/417bfaca60ef103dec3e342d41184c547e2f2604)

@MechCoder , since `_init_centroids ` will always be called after the code block you mentioned and thus make sure centers has the same `dtype` as input, I think maybe we don't need to touch that part of code?